daemon/api.go

```diff
@@ -1082,16 +1085,37 @@ out:
 		origPath = form.Value["snap-path"][0]
 	}
 
-	info, err := readSnapInfo(tempPath)
-	if err != nil {
-		return InternalError("cannot read snap file: %v", err)
-	}
-	snapName := info.Name()
-
 	st := c.d.overlord.State()
 	st.Lock()
 	defer st.Unlock()
 
+	var snapName string
```

Comment:
I wonder if it would make sense to move this code block into a small helper to keep postSnap from growing too much, something like `func infoFromFile(st state.State, path string, dangerousOk bool) (name string, si snap.SideInfo, err error)`. But maybe overkill (given that the error handling is also slightly different).